src/Prefetch.cpp

```diff
@@ -3,214 +3,200 @@
-
-        if (prefetches) {
-            for (const Prefetch &p : *prefetches) {
-                if (!ends_with(op->name, "." + p.var)) {
+        scope.pop(op->name);
+
+        if (!prefetch_list.empty()) {
+            // If there are multiple prefetches of the same Func or ImageParam,
+            // use the most recent one
+            set<string> seen;
+            for (int i = prefetch_list.size() - 1; i >= 0; --i) {
+                const PrefetchDirective &p = prefetch_list[i];
+                if (!ends_with(op->name, "." + p.var) || (seen.find(p.name) != seen.end())) {
                     continue;
                 }
+                seen.insert(p.name);
+
                 // Add loop variable + prefetch offset to interval scope for box computation
                 Expr fetch_at = loop_var + p.offset;
-                bounds.push(op->name, Interval(fetch_at, fetch_at));
-                map<string, Box> boxes_read = boxes_required(body, bounds);
-                bounds.pop(op->name);
-
-                // Don't prefetch buffers that are written to. We assume that these already
-                // have good locality.
-                // TODO: This is not a good assumption. It would be better to have the
-                // prefetch directive specify the buffer that we want to prefetch, instead
-                // of trying to figure out which buffers should be prefetched. This would also
-                // mean that we don't need the "make_similar_load" hack, because we can make
-                // calls the standard way (using the ImageParam/Function object referenced in
-                // the prefetch).
-                map<string, Box> boxes_written = boxes_provided(body, bounds);
-                for (const auto &b : boxes_written) {
-                    auto it = boxes_read.find(b.first);
-                    if (it != boxes_read.end()) {
-                        debug(2) << "Not prefetching buffer " << it->first
-                                 << " also written in loop " << op->name << "\n";
-                        boxes_read.erase(it);
-                    }
-                }
-
-                // TODO: Only prefetch the newly accessed data from the previous iteration.
-                // This should use boxes_touched (instead of boxes_required) so we exclude memory
-                // either read or written.
-                for (const auto &b : boxes_read) {
-                    const string &buf_name = b.first;
-
+                scope.push(op->name, Interval(fetch_at, fetch_at));
+                map<string, Box> boxes_rw = boxes_touched(body, scope);
+                scope.pop(op->name);
+
+                // TODO(psuriana): Only prefetch the newly accessed data. We
+                // should subtract the box accessed during previous iteration
+                // from the one accessed during this iteration.
+
+                // TODO(psuriana): Shift the base address of the prefetched box
```

Comment:
This TODO should be updated. I think it now should read "Add a new PrefetchBoundStrategy::ShiftInwards that shifts the min..."